daml/Marketplace/Distribution/Auction/Service.daml

```diff
@@ -0,0 +1,201 @@
+        archive bidServiceCid
+
+    nonconsuming choice AddLeadBidder : ContractId LeadBidder
+      with
+        submitter : Party
+        leadBidCid : ContractId Bid.T
+      controller submitter
+      do
+        assertMsg ("Unauthorised Submitter - " <> show submitter) (submitter `elem` [provider, customer])
+        bid <- fetch leadBidCid
+        (_, auction) <- fetchByKey @Auction (bid.provider, bid.assetId)
+        assertMsg ("Unexpected lead bidder - " <> show bid.customer) (bid.customer == auction.leadBidder)
+
+        let
+          allocationPercentage = roundBankers 2 $ bid.details.quantity / auction.asset.quantity
+          leadBidder = bid.customer
+        create LeadBidder with ..
+
+    controller customer can
+      nonconsuming RequestCreateAuction : ContractId CreateAuctionRequest
+        with
+          leadBidder : Party
+          asset : Asset
+          quotedAssetId : Id
+          assetDepositCid : ContractId AssetDeposit
+          allocationAccount : Account
+          receivableAccount : Account
+          floorPrice : Decimal
+        do
+          create CreateAuctionRequest with ..
+
+      nonconsuming CancelAuctionRequest : ()
+        with
+          createAuctionRequestCid : ContractId CreateAuctionRequest
+        do
+          archive createAuctionRequestCid
+
+    controller provider can
+      nonconsuming CreateAuction : ContractId Auction.T
+        with
+          createAuctionRequestCid : ContractId CreateAuctionRequest
+        do
+          CreateAuctionRequest{..} <- fetchAndArchive createAuctionRequestCid
+
+          -- Deposit the asset into the allocation account so they cannot be transferred/withdrawn without the signatore of the Auctioneer
+          depositCid <- exerciseByKey @AllocationAccountRule allocationAccount.id Allocation.Deposit with depositCid = assetDepositCid; ..
+          create Auction with status = Auction.Open; ..
+
+      nonconsuming RejectAuction : ()
+        with
+          createAuctionRequestCid : ContractId CreateAuctionRequest
+        do
+          archive createAuctionRequestCid
+
+      nonconsuming ProcessAuction : (ContractId Auction.T, [ContractId Settlement.SettlementInstruction])
+        with
+          auctionCid : ContractId Auction.T
+          bidCids : [ContractId Bid.T]
+        do
+          auction@Auction{..} <- fetch auctionCid
+          assertMsg "Auction already successfully processed" (status `elem` [Auction.Open, Auction.NoValidBids])
+          bids <- forA bidCids fetch
+
+          -- Split bids on those above / below the floor price -> Sort valid bids descendingly -> Execute Auction -> Split auction result on allocated / unallocated (ie, zero allocation)
+          let
+            (validBids, invalidBids) = splitList (\bid -> bid.details.price < floorPrice) bids
+            sortedBids = sortOn (\bid -> Down bid.details) validBids
```

Comment:
I'd maybe make this choice a wrapper on a more general function, taking a type constraint `Ord t` and constructor `(* -> t)`, so that this can be passed in by the user instead of `Down` you have here. i.e. the user can choose what ordering they want.